### PR TITLE
6201035: Document NPE for passing null insets to constructors and methods of several javax.swing.border.* classes

### PR DESCRIPTION
**Pull request by prsadhuk**
The behavior of MatteBorder constructors taking Insets object as a parameter is undocumented. It would throw NPE if null object is passed to it, which should be documented in the spec.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-6201035](https://bugs.openjdk.org/browse/JDK-6201035): Document NPE for passing null insets to constructors and methods of several javax.swing.border.* classes
 * [JDK-8295454](https://bugs.openjdk.org/browse/JDK-8295454): Document NPE for passing null insets to constructors and methods of several javax.swing.border.* classes (**CSR**)

### Reviewers
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Committer) ⚠️ Review applies to [389c7c4c](https://git.openjdk.org/jdk/pull/10740/files/389c7c4c69c7a5e5b016a82b9dc552a85f4ed3a5)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [389c7c4c](https://git.openjdk.org/jdk/pull/10740/files/389c7c4c69c7a5e5b016a82b9dc552a85f4ed3a5)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10740/head:pull/10740` \
`$ git checkout pull/10740`

Update a local copy of the PR: \
`$ git checkout pull/10740` \
`$ git pull https://git.openjdk.org/jdk pull/10740/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10740`

View PR using the GUI difftool: \
`$ git pr show -t 10740`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10740.diff">https://git.openjdk.org/jdk/pull/10740.diff</a>

</details>

### CODE REVIEW

**Comment by mrserb** [comment]
It seems the same issue exists in other methods, like getBorderInsets?

**Comment by prsadhuk** [comment]
> It seems the same issue exists in other methods, like getBorderInsets?

OK. javadoc updated for that method too

File: src/java.desktop/share/classes/javax/swing/border/MatteBorder.java [hunk 1]
```diff
@@ -169,7 +169,9 @@ private void paintEdge(Component c, Graphics g, int x, int y, int width, int hei
     /**
      * Reinitialize the insets parameter with this Border's current Insets.
      * @param c the component for which this border insets value applies
-     * @param insets the object to be reinitialized
+     * @param insets the object to be
```
[the remark this answers is not in the document]
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/border/MatteBorder.java]: yes..rectified.

**Review by TejeshR13 — APPROVED** [review]

File: src/java.desktop/share/classes/javax/swing/border/MatteBorder.java [hunk 2]
```diff
@@ -75,6 +75,8 @@ public MatteBorder(int top, int left, int bottom, int right, Color matteColor)
      * Creates a matte border with the specified insets and color.
      * @param borderInsets the insets of the border
      * @param matteColor the color rendered for the border
+     * @throws {@code NullPointerException} if {@code borderInsets}
+     *          is {@code null}
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/MatteBorder.java:79]: Would `@inheritDoc` work in this case? The message is the same as in its super class.
```suggestion
     * @throws {@inheritDoc}
```

Comment (mrserb) [reply src/java.desktop/share/classes/javax/swing/border/MatteBorder.java:79]: If we updated the parent class should not we do the same for all upto AbstractBorder?

Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/border/MatteBorder.java:79]: I guess we should..updated..

File: src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java [hunk 3]
```diff
@@ -85,6 +85,8 @@ public Insets getBorderInsets(Component c)       {
      * @param c the component for which this border insets value applies
      * @param insets the object to be reinitialized
      * @return the <code>insets</code> object
+     * @throws {@code NullPointerException} if {@code insets}
+     *         is {@code null}
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java:89]: This does not compile, `@throws` expects a “plain” class name, monospaced font is applied automatically to `NullPointerException`.

File: src/java.desktop/share/classes/javax/swing/border/EmptyBorder.java [hunk 4]
```diff
@@ -102,6 +104,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      * Reinitialize the insets parameter with this Border's current Insets.
      * @param c the component for which this border insets value applies
      * @param insets the object to be reinitialized
+     * @throws {@code NullPointerException} if {@code insets}
+     *          is {@code null}
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/EmptyBorder.java:108]:
```suggestion
     * @throws NullPointerException {@inheritDoc}
```
I suggest using `{@inheritDoc}` to copy the exception description from the super class for both `EmptyBorder` and `MatteBorder`. (It cannot be used for constructors though.)

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]
In addition to the comments in the code, could you please update the copyright year and remove `Rectangle` from imports in `EmptyBorder`?

**Comment by aivanov-jdk** [comment]
Does it make sense to update the subject to _“Document NPE for passing null insets”_ or a similar? It would carry more meaning. At this time, “underfined behavior” sounds frightening.

File: src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java [hunk 5]
```diff
@@ -85,6 +85,8 @@ public Insets getBorderInsets(Component c)       {
...
      */
     public Insets getBorderInsets(Component c, Insets insets) {
```
[the remark this answers is not in the document]
Comment (mrserb) [reply src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java:91]: Did we cover all the subclasses?

[the remark this answers is not in the document]
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java:91]: More classes added

**Comment by prsadhuk** [comment]
Changed subject and added more border classes/methods

File: src/java.desktop/share/classes/javax/swing/border/BevelBorder.java [hunk 6]
```diff
@@ -142,6 +142,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      * Reinitialize the insets parameter with this Border's current Insets.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/BevelBorder.java:142]: To be consistent with the superclass:
```suggestion
     * Reinitializes the insets parameter with this Border's current Insets.
```

I guess the entire javadoc can be inherited with {@inheritDoc}, it adds nothing to the text in `AbstractBorder`.

File: src/java.desktop/share/classes/javax/swing/border/CompoundBorder.java [hunk 7]
```diff
@@ -137,6 +137,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      * Reinitialize the insets parameter with this Border's current Insets.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/CompoundBorder.java:137]:
```suggestion
     * Reinitializes the insets parameter with this Border's current Insets.
```
The same: the verb at start of a javadoc is in third person singular.

File: src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java [hunk 8]
```diff
@@ -85,6 +85,8 @@ public Insets getBorderInsets(Component c)       {
      * @param c the component for which this border insets value applies
      * @param insets the object to be reinitialized
      * @return the <code>insets</code> object
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/AbstractBorder.java:87]:
```suggestion
     * @return the {@code insets} object
```
It makes sense to replace `<code>` with `{@code}` as the `throws` clause uses the new syntax.

File: src/java.desktop/share/classes/javax/swing/border/EmptyBorder.java [hunk 9]
```diff
@@ -102,6 +104,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      * Reinitialize the insets parameter with this Border's current Insets.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/EmptyBorder.java:104]:
```suggestion
     * Reinitializes the insets parameter with this Border's current Insets.
```

File: src/java.desktop/share/classes/javax/swing/border/EtchedBorder.java [hunk 10]
```diff
@@ -210,6 +210,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      *
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/EtchedBorder.java:210]: Also add _“-s”_ ending to the verb in the line above.

File: src/java.desktop/share/classes/javax/swing/border/LineBorder.java [hunk 11]
```diff
@@ -174,6 +174,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      *
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/LineBorder.java:174]: Also add _“-s”_ ending to the verb in the line above.

File: src/java.desktop/share/classes/javax/swing/border/SoftBevelBorder.java [hunk 12]
```diff
@@ -152,6 +152,8 @@ public void paintBorder(Component c, Graphics g, int x, int y, int width, int he
      * Reinitialize the insets parameter with this Border's current Insets.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/border/SoftBevelBorder.java:152]:
```suggestion
     * Reinitializes the insets parameter with this Border's current Insets.
```

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]
`TitledBorder` is to be updated as well.

I wonder why none of the classes uses `{@inheritDoc}` instead of copying the entire text without modification. And you don't want to use `{@inheritDoc}` either for the `@throws` clause without explaining why.

**Comment by aivanov-jdk** [comment]
> `TitledBorder` is to be updates as well.

https://github.com/openjdk/jdk/blob/1a58cb1c023c876594e8a53d00703e564a922d36/src/java.desktop/share/classes/javax/swing/border/TitledBorder.java#L373-L378

This hasn't been addressed yet. It requires the CSR to be updated.

**Comment by aivanov-jdk** [comment]
Before I can approve this PR, I would like to see the javadoc for border classes (those that are modified) updated so that all of them use *“Reinitializes”* as the first word in the description instead of *“Reinitialize”*, which aligns with [the javadoc style guide](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#styleguide):

> **Use 3rd person (descriptive) not 2nd person (prescriptive).**
> 
> The description is in 3rd person declarative rather than 2nd person imperative.
>
> Gets the label. *(preferred)*
>
> Get the label. (avoid)

**Review by aivanov-jdk — APPROVED** [review]

**Comment by aivanov-jdk** [comment]
> Before I can approve this PR, I would like to see the javadoc for border classes (those that are modified) updated so that all of them use _“Reinitializes”_ as the first word in the description instead of _“Reinitialize”_, which aligns with [the javadoc style guide](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#styleguide):
> 
> > **Use 3rd person (descriptive) not 2nd person (prescriptive).**
> > The description is in 3rd person declarative rather than 2nd person imperative.
> > Gets the label. _(preferred)_
> > Get the label. (avoid)

Per the comment in JBS, addressing this is deferred to a separate issue… However, I still think it would've been more efficient to handle it here. Anyway, it's approved now.

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**